sycl/plugins/xrt/pi_xrt.cpp

```diff
@@ -0,0 +1,2312 @@
+  value->decrement_reference_count();
+}
+template <typename T, typename std::enable_if_t<
+                          !std::is_base_of_v<ref_counted_base<T>, T>, int> = 0>
+void decr_ref_count(T *value) {}
+
+template <typename T, typename std::enable_if_t<
+                          std::is_base_of_v<ref_counted_base<T>, T>, int> = 0>
+void assert_valid_obj(T *obj) {
+  assert(obj);
+  assert(obj->get_reference_count() > 0);
+}
+
+template <typename T, typename std::enable_if_t<
+                          !std::is_base_of_v<ref_counted_base<T>, T>, int> = 0>
+void assert_valid_obj(T *obj) {
+  assert(obj);
+}
+
+template <typename T> void assert_valid_objs(T ptr, int count) {
+  for (int i = 0; i < count; i++)
+    assert_valid_obj(ptr[i]);
+}
+
+/// wrapper type for a _pi_* object that keeps a reference on an other _pi_*
+/// object that might be refcounted
+template <typename T> struct ref_counted_ref {
+  T *value = nullptr;
+  ref_counted_ref() = default;
+  ref_counted_ref(std::nullptr_t) : ref_counted_ref() {}
+
+  /// Copy from an existing ref so we increase the ref count
+  ref_counted_ref(T *v) : value(v) { retain(); }
+  ref_counted_ref(const ref_counted_ref &other)
+      : ref_counted_ref(other.value) {}
+  ref_counted_ref(ref_counted_ref &&other) : value(other.value) {
+    other.value = nullptr;
+  }
+
+  T *operator->() { return value; }
+  T &operator*() { return *value; }
+  operator bool() { return value; }
+  T *get() { return value; }
+
+  /// to be used when providing a raw pointer to the user of the PI
+  T *give_externally() {
+    retain();
+    return value;
+  }
+
+  void release() {
+    if (value)
+      decr_ref_count(value);
+  }
+  void retain() {
+    if (value)
+      incr_ref_count(value);
+  }
+  void reset() {
+    release();
+    value = nullptr;
+  }
+
+  void swap(ref_counted_ref &other) {
+    T *tmp = value;
+    value = other.value;
+    other.value = tmp;
+  }
+
+  ref_counted_ref &operator=(ref_counted_ref other) {
+    swap(other);
+    return *this;
+  }
+  ~ref_counted_ref() { release(); }
+};
+
+template <typename base, typename... Args>
+ref_counted_ref<base> make_ref_counted(Args &&...args) {
+  return new base(std::forward<Args>(args)...);
+}
+
+template<typename T>
+struct singleton {
+private:
+  /// Storage for the instance
+  static T *&get_self() {
+    static T *self = nullptr;
+    return self;
+  }
+public:
+  /// Return a new instance or the currently live instance
+  template <typename... Ts> static ref_counted_ref<T> get(Ts &&...ts) {
+    if (!get_self()) {
+      auto res = make_ref_counted<T>(std::forward<Ts>(ts)...);
+      get_self() = res.get();
+      return res;
+    }
+    return get_self();
+  }
+  ~singleton() {
+    /// Make null such that the next request will rebuild
+    get_self() = nullptr;
+  }
+};
+
+/// Cannot be used for XRT objects because xrt also has global destructors
+struct cleanup_system {
+  using clenup_item = std::pair<void (*)(void *), void *>;
```

Comment:
```suggestion
  using cleanup_item = std::pair<void (*)(void *), void *>;
```
but honestly, rather to add this alias, create a `struct` with 2 sensible member names instead of `first` and `second`.